timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,287 @@
+        all_frames.append(frames)
+        all_boxes.append(boxes)
+
+    return all_frames, all_boxes, all_U_fns
+
+
+def plot_BAR(dG, dG_err, fwd_delta_u, rev_delta_u, title, plot_idx):
+    plt.subplot(*plot_idx)
+    plt.title(f"{title}, dG: {dG:.2f} +- {dG_err:.2f} kTs")
+    plt.hist(fwd_delta_u, alpha=0.5, label="fwd", density=True)
+    plt.hist(-rev_delta_u, alpha=0.5, label="-rev", density=True)
+    plt.xlabel("work (kTs)")
+    plt.legend()
+
+
+SimulationResult = namedtuple("SimulationResult", ["all_dGs", "all_errs", "plot_png", "frames", "boxes", "U_fns"])
+
+
+def estimate_free_energy_given_samples(all_frames, all_boxes, all_U_fns, lambda_schedule, prefix):
+    # assume pair-BAR format
+    # setup U_fns to operate in batch mode
+    kT = BOLTZ * 300.0
+    beta = 1 / kT
+
+    all_dGs = []
+    all_errs = []
+
+    all_batch_U_fns = []
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        bound_impls = [p.bound_impl(np.float32) for p in all_U_fns[lamb_idx]]
+        batch_U_fns = get_batch_U_fns(bound_impls, lamb)
+        all_batch_U_fns.append(batch_U_fns)
+
+    U_names = []
+    for U_fn in all_U_fns[0]:
+        # convert from '<timemachine.lib.potentials.Nonbonded object at 0x7f7880b900b8>' -> Nonbonded
+        U_names.append(repr(U_fn).split(".")[-1].split()[0])
+
+    num_rows = len(lambda_schedule) - 1
+    num_cols = len(U_names) + 1
+
+    plot_iota = 1
+
+    def get_plot_idx():
+        nonlocal plot_iota  # what is this black magic? needed to deal with += increment
+        idx = (num_rows, num_cols, plot_iota)
+        plot_iota += 1
+        return idx
+
+    plt.figure(figsize=(num_cols * 5, num_rows * 3))
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        if lamb_idx > 0:
+
+            prev_frames = all_frames[lamb_idx - 1]
+            cur_frames = all_frames[lamb_idx]
+
+            prev_boxes = all_boxes[lamb_idx - 1]
+            cur_boxes = all_boxes[lamb_idx]
+
+            prev_U_fns = all_batch_U_fns[lamb_idx - 1]
+            cur_U_fns = all_batch_U_fns[lamb_idx]
+
+            # loop over bond, angle, torsion, nonbonded terms etc.
+            all_fwd_delta_us = []
+            all_rev_delta_us = []
+
+            for u_idx, (prev_U_fn, cur_U_fn) in enumerate(zip(prev_U_fns, cur_U_fns)):
+                fwd_delta_u = beta * (cur_U_fn(prev_frames, prev_boxes) - prev_U_fn(prev_frames, prev_boxes))
+                rev_delta_u = beta * (prev_U_fn(cur_frames, cur_boxes) - cur_U_fn(cur_frames, cur_boxes))
+
+                dG, dG_err = pymbar.BAR(fwd_delta_u, rev_delta_u)
+                dG /= beta
+                dG_err /= beta
+
+                plot_BAR(dG, dG_err, fwd_delta_u, rev_delta_u, U_names[u_idx], get_plot_idx())
+
+                all_fwd_delta_us.append(fwd_delta_u)
+                all_rev_delta_us.append(rev_delta_u)
+
+            # sanity check - I don't think the dG calculation commutes with its components, so we have to re-estimate
```

Comment:
@maxentile this is where the delta_U decomposition is done, I've added a comment here for future readers